File: src/com/fenghuo/notes/alarm/MyAlarm.java
package com.fenghuo.notes.alarm;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;

import com.fenghuo.notes.db.DBAlarmHelper;
import com.fenghuo.notes.db.DBNoteHelper;
import com.fenghuo.notes.bean.Alarm;

public class MyAlarm {

	private AlarmManager manager;
	private Context context;
	private DBAlarmHelper alarmHelper;
	private DBNoteHelper noteHelper;
	private Date datenow;

	public MyAlarm(Context context) {
		this.context = context;
		manager = (AlarmManager) context
				.getSystemService(Context.ALARM_SERVICE);
		alarmHelper = new DBAlarmHelper(context);
		noteHelper = new DBNoteHelper(context);
	}

	/**
	 * 设置闹钟
	 *
	 * @param alarm
	 */
	public void set(Alarm alarm) {

		// 当前的时间
		datenow = new Date();
		// 重复的星期
		String[] weekstr = null;
		int[] weekint = null;
		if (alarm.getWeek().length() > 0) {
			weekstr = alarm.getWeek().split(",");
			weekint = new int[weekstr.length];
			for (int i = 0; i < weekstr.length; i++) {
				weekint[i] = Integer.parseInt(weekstr[i]);
			}
		}

		// 闹钟时间
		SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm");
		Date date = null;
		try {
			date = dateFormat.parse(alarm.getDate() + " " + alarm.getTime());
		} catch (ParseException e) {
			e.printStackTrace();
		}

		//判断是否过时(当天)
		Date datetimenow=new Date(2000, 1, 1, datenow.getHours(), datenow.getMinutes());
		Date datetimering=new Date(2000, 1, 1, date.getHours(), date.getMinutes());
		System.out.println(datetimenow.toLocaleString()+"\n"+datetimering.toLocaleString());
		if(datetimenow.after(datetimering)){
			return;
		}

		Intent intent = new Intent();
		intent.putExtra("alarmid", alarm.getId());
		intent.setAction(AlarmReciver.Action_StartAlarm);
		PendingIntent pendingIntent = PendingIntent.getBroadcast(context, alarm.getId(),intent, PendingIntent.FLAG_UPDATE_CURRENT);


		// 有重复
		if (weekint != null && weekint.length > 0) {
			int today = datenow.getDay() + 1;
			boolean have = false;
			for (int i = 0; i < weekint.length; i++) {
				if (weekint[i] == today)
					have = true;
			}
			if (have) {
				Calendar calendar = Calendar.getInstance();
				calendar.setTime(datenow);
				calendar.set(Calendar.HOUR_OF_DAY, date.getHours());
				calendar.set(Calendar.MINUTE, date.getMinutes());
				manager.cancel(pendingIntent);
				manager.set(AlarmManager.RTC_WAKEUP,
						calendar.getTimeInMillis(), pendingIntent);
			}
		}
		// 无重复

		PendingIntent pendingIntent2 = PendingIntent.getBroadcast(context, alarm.getId(),
				intent, PendingIntent.FLAG_UPDATE_CURRENT);
		Calendar calendar = Calendar.getInstance();
		calendar.setTime(date);
		manager.cancel(pendingIntent2);
		manager.set(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis(),
				pendingIntent2);
	}

	/**
	 * 删除一个闹钟
	 *
	 * @param id
	 */
	public void delete(int id) {
		// 从系统闹钟队列中删除
		Intent intent = new Intent();
		intent.setAction(AlarmReciver.Action_StartAlarm);
		PendingIntent pendingIntent = PendingIntent.getBroadcast(context, 1,
				intent, 0x102 + id);
		PendingIntent pendingIntent2 = PendingIntent.getBroadcast(context, 1,
				intent, 0x503 + id);
		manager.cancel(pendingIntent);
		manager.cancel(pendingIntent2);
	}

	/**
	 * 开机以后重新设置
	 */
	public void reset() {
		List<Alarm> alarms = alarmHelper.Getall();
		for (Alarm alarm : alarms) {
			set(alarm);
		}
	}

	/**
	 * 使用后必须调用 用来释放数据库连接
	 */
	public void destroy() {
		alarmHelper.Destroy();
		noteHelper.Desdroy();
	}

}

File: src/com/fenghuo/notes/alarm/MyAlarm.kt
package com.fenghuo.notes.alarm

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import com.fenghuo.notes.bean.Alarm
import com.fenghuo.notes.db.DBAlarmHelper
import com.fenghuo.notes.db.DBNoteHelper
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

class MyAlarm(private val context: Context) {

    private val manager: AlarmManager
    private val alarmHelper: DBAlarmHelper
    private val noteHelper: DBNoteHelper
    private var datenow: Date? = null

    init {
        manager = context
                .getSystemService(Context.ALARM_SERVICE) as AlarmManager
        alarmHelper = DBAlarmHelper(context)
        noteHelper = DBNoteHelper(context)
    }

    /**
     * 设置闹钟

     * @param alarm
     */
    fun set(alarm: Alarm) {

        // 当前的时间
        datenow = Date()
        // 重复的星期
        var weekstr: Array<String>? = null
        var weekint: IntArray? = null
        var week: String? = alarm.week
        if (week != null && week.isNotEmpty()) {
            weekstr = alarm.week!!.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            weekint = IntArray(weekstr.size)
            for (i in weekstr.indices) {
                weekint[i] = Integer.parseInt(weekstr[i])
            }
        }

        // 闹钟时间
        val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm")
        var date: Date? = null
        try {
            date = dateFormat.parse(alarm.date + " " + alarm.time)
        } catch (e: ParseException) {
            e.printStackTrace()
        }

        //判断是否过时(当天)
        val datetimenow = Date(2000, 1, 1, datenow!!.hours, datenow!!.minutes)
        val datetimering = Date(2000, 1, 1, date!!.hours, date.minutes)
        println(datetimenow.toLocaleString() + "\n" + datetimering.toLocaleString())
        if (datetimenow.after(datetimering)) {
            return
        }

        val intent = Intent()
        intent.putExtra("alarmid", alarm.id)
        intent.action = AlarmReciver.Action_StartAlarm
        val pendingIntent = PendingIntent.getBroadcast(context, alarm.id, intent, PendingIntent.FLAG_UPDATE_CURRENT)


        // 有重复
        if (weekint != null && weekint.size > 0) {
            val today = datenow!!.day + 1
            var have = false
            for (i in weekint.indices) {
                if (weekint[i] == today)
                    have = true
            }
            if (have) {
                val calendar = Calendar.getInstance()
                calendar.time = datenow
                calendar.set(Calendar.HOUR_OF_DAY, date.hours)
                calendar.set(Calendar.MINUTE, date.minutes)
                manager.cancel(pendingIntent)
                manager.set(AlarmManager.RTC_WAKEUP,
                        calendar.timeInMillis, pendingIntent)
            }
        }
        // 无重复

        val pendingIntent2 = PendingIntent.getBroadcast(context, alarm.id,
                intent, PendingIntent.FLAG_UPDATE_CURRENT)
        val calendar = Calendar.getInstance()
        calendar.time = date
        manager.cancel(pendingIntent2)
        manager.set(AlarmManager.RTC_WAKEUP, calendar.timeInMillis,
                pendingIntent2)
    }

    /**
     * 删除一个闹钟

     * @param id
     */
    fun delete(id: Int) {
        // 从系统闹钟队列中删除
        val intent = Intent()
        intent.action = AlarmReciver.Action_StartAlarm
        val pendingIntent = PendingIntent.getBroadcast(context, 1, intent, 0x102 + id)
        val pendingIntent2 = PendingIntent.getBroadcast(context, 1, intent, 0x503 + id)
        manager.cancel(pendingIntent)
        manager.cancel(pendingIntent2)
    }

    /**
     * 开机以后重新设置
     */
    fun reset() {
        val alarms = alarmHelper.Getall()
        for (alarm in alarms) {
            set(alarm)
        }
    }

    /**
     * 使用后必须调用 用来释放数据库连接
     */
    fun destroy() {
        alarmHelper.Destroy()
        noteHelper.Desdroy()
    }

}
